### DISCUSSION TITLE
How to use microtime() format in Admin UI - is it possible?

### DISCUSSION THREAD

**Discussion by Jimmi08** [e107inc/e107#5195] [Q&A Core] [answered]
For now, I use an integer field   with microtime(true)  value to save time as an integer value (saving actual time)   

For display I plan to use method and display time in readable form - only hour, minutes, seconds and thousands of seconds (like for some race timing). 

But I will need to edit this field too. 

Is there other way? Somehow to use datetime type field?  

I think that this is not supported in e107, at least I haven't find anything in code. But maybe I am mistaken.    

Thank you for any help

**Comment by Deltik** [comment]
There is an example implementation of a datetime picker in e107 at `/e107_plugins/download/admin_download.php?mode=main&action=create`:
![`/e107_plugins/download/admin_download.php?mode=main&action=create`](https://github.com/e107inc/e107/assets/1364268/3d3e5813-de20-42d2-a561-019b777ffc42)

The method to use is `e_form::datepicker()`, and you can find an example usage [in `download_main_admin_ui::create_download()`](https://github.com/e107inc/e107/blob/8921e92aa62cf57a5fc76bb5bd11af210541de13/e107_plugins/download/includes/admin.php#L1558-L1558)

**Reply by Jimmi08** [reply]
Thank you, but it is formatted - data and time in hours and seconds.    I didn't find examples of how to use milliseconds.    

Example what I need:
- start race is 14:00:00.0000 
- first control  is 14:02:02.5672

Saving is not the problem, it is just microtome(true) instead time().  
Displaying and editing I am able to solve with method type.

I just need the confirmation that this is possible to solve in e107 datetime type, I will dig in formats then.  I know that seconds work, but what about milliseconds?

**Reply by Deltik** [reply, marked as answer]
Ah, in this case, I would recommend creating your own input field and converting between the Unix epoch and the ISO 8601 time (which looks like `2024-02-14T08:33:28.8829+00:00`).  You can then change the milliseconds, and when saving, convert the ISO 8601 time back to the database representation (e.g. `1707899608.8829`).

Example functions that can convert between the two:

```php
/**
 * Converts a microtime float to an ISO 8601 datetime string.
 *
 * @param float $microtime The microtime float to convert.
 * @return string The ISO 8601 datetime string.
 */
function microtimeToISO8601(float $microtime): string
{
    $seconds = floor($microtime);
    $microseconds = $microtime - $seconds;
    $datetime = new DateTime("@$seconds");
    $iso8601 = $datetime->format('Y-m-d\TH:i:s');
    $iso8601 .= rtrim(sprintf('.%06d', $microseconds * 1000000), '0');
    $iso8601 .= $datetime->format('P');
    return $iso8601;
}

/**
 * Converts an ISO 8601 datetime string to a microtime float.
 *
 * @param string $iso8601 The ISO 8601 datetime string to convert.
 * @return float The microtime float.
 */
function ISO8601ToMicrotime(string $iso8601): float
{
    $datetime = DateTime::createFromFormat('Y-m-d\TH:i:s.uP', $iso8601);
    if ($datetime === false) {
        throw new InvalidArgumentException('Invalid ISO 8601 datetime string.');
    }
    $microtime = (float) $datetime->format('U.u');
    return $microtime;
}
```

**Reply by Jimmi08** [reply]
Yes, I thought that.  I just wanted to be sure not invent a wheel again.  Thank you very much for the examples.  They look differently as I have found already, I need to investigate this more.  This is big help,

**Reply by Jimmi08** [reply]
Just info about progress.

Needed format: 13:16:03.082 
Most of the records are from RFID readers in this format, some are manual and some will need editing.  So it would be saved as text, it is easier that way. 

I am not sure if the e107 environment has influenced this, but the e107 timezone is +1 hour.

I had to make those changes (no idea why, just trying and checking result) : 

- `$datetime = new DateTime(); `-  removed"@$seconds" - otherwise time was one hour back
- changed to microseconds (u versus v) 
- removed `format('P')`
- changed format, to have only time

Result (working for now):

```
 
function microtimeToISO8601(float $microtime): string
{
    $seconds = floor($microtime);
    $microseconds = $microtime - $seconds;
    $miliseconds = round($microseconds * 1000, 0);
    $datetime = new DateTime();
    $iso8601 = $datetime->format('H:i:s');
    $iso8601 .= rtrim(sprintf('.%03d', $miliseconds ), '0');
    //$iso8601 .= $datetime->format('P');
    return $iso8601;
}

 
function ISO8601ToMicrotime(string $iso8601): float
{
    $datetime = DateTime::createFromFormat('H:i:s.v', $iso8601);
    if ($datetime === false)
    {
        throw new InvalidArgumentException('Invalid ISO 8601 datetime string.');
    }

    $militime = (float) $datetime->format('U.v');
    return $militime;
}
```
